node/cn/peer_set.go

```diff
@@ -469,3 +510,67 @@ func (peers *peerSet) UpdateTypePeersWithoutTxs(tx *types.Transaction, nodeType
 	}
 	logger.Trace("Broadcast transaction", "hash", tx.Hash(), "recipients", len(typePeers))
 }
+
+// RegisterSnapExtension unblocks an already connected `klay` peer waiting for its
+// `snap` extension, or if no such peer exists, tracks the extension for the time
+// being until the `eth` main protocol starts looking for it.
+func (peers *peerSet) RegisterSnapExtension(peer *snap.Peer) error {
+	// Reject the peer if it advertises `snap` without `klay` as `snap` is only a
+	// satellite protocol meaningful with the chain selection of `klay`
+	if !peer.RunningCap(backend.IstanbulProtocol.Name, backend.IstanbulProtocol.Versions) {
+		return errSnapWithoutIstanbul
+	}
+	// Ensure nobody can double connect
+	peers.lock.Lock()
+	defer peers.lock.Unlock()
+
+	id := peer.ID()
+	if _, ok := peers.peers[id]; ok {
+		return errPeerAlreadyRegistered // avoid connections with the same id as existing ones
+	}
+	if _, ok := peers.snapPend[id]; ok {
+		return errPeerAlreadyRegistered // avoid connections with the same id as pending ones
+	}
+	// Inject the peer into an `eth` counterpart is available, otherwise save for later
+	if wait, ok := peers.snapWait[id]; ok {
+		delete(peers.snapWait, id)
+		wait <- peer
+		return nil
+	}
+	peers.snapPend[id] = peer
+	return nil
+}
+
+// WaitSnapExtension blocks until all satellite protocols are connected and tracked
+// by the peerset.
+func (ps *peerSet) WaitSnapExtension(peer Peer) (*snap.Peer, error) {
+	// If the peer does not support a compatible `snap`, don't wait
+	if !peer.RunningCap(snap.ProtocolName, snap.ProtocolVersions) {
+		return nil, nil
+	}
+	// Ensure nobody can double connect
+	ps.lock.Lock()
+
```

Comment:
It would be good to make an internal function and use `defer` for better code maintenance.

Reply:
I have refactored this on 154e3e65c76c7d0b50f7c35b431768505ceb6104. Please take another look.